tiledb/cloud/array.py

```diff
@@ -182,3 +186,177 @@ def array_activity(uri):
         return api_instance.array_activity_log(namespace=namespace, array=array_name)
     except GenApiException as exc:
         raise tiledb_cloud_error.check_exc(exc) from None
+
+
+def split_uri(uri):
+    """
+  Split a URI into namespace and array name
+
+  :param uri: uri to split into namespace and array name
+  :return: tuple (namespace, array_name)
+  """
+    parsed = urllib.parse.urlparse(uri)
+    if not parsed.scheme == "tiledb":
+        raise Exception("Incorrect array uri, must be in tiledb:// scheme")
+    return parsed.netloc, parsed.path[1:]
+
+
+def parse_ranges(ranges):
+    """
+    Takes a list of the following objects per dimension:
+
+    - scalar index
+    - (start,end) tuple
+    - list of either of the above types
+
+    :param ranges: list of (scalar, tuple, list)
+    :param builder: function taking arguments (dim_idx, start, end)
+    :return:
+    """
+
+    def make_range(dim_range):
+        if isinstance(dim_range, (int, float)):
+            start, end = dim_range, dim_range
+        elif isinstance(dim_range, (tuple, list)):
+            start, end = dim_range[0], dim_range[1]
+        elif isinstance(dim_range, slice):
+            assert dim_range.step is None, "slice steps are not supported!"
+            start, end = dim_range.start, dim_range.stop
+        else:
+            raise ValueError("Unknown index type! (type: '{}')".format(type(dim_range)))
+        return [start, end]
+
+    result = list()
+    for dim_idx, dim_range in enumerate(ranges):
+        dim_list = []
+        # TODO handle numpy scalars here?
+        if isinstance(dim_range, (int, float, tuple, slice)):
+            dim_list.extend(make_range(dim_range))
+        elif isinstance(dim_range, list):
+            for r in dim_range:
+                dim_list.extend(make_range(r))
+        else:
+            raise ValueError(
+                "Unknown subarray/index type! (type: '{}', "
+                ", idx: '{}', value: '{}')".format(type(dim_range), dim_idx, dim_range)
+            )
+        result.append(dim_list)
+
+    return result
+
+
+def apply_async(
+    uri,
+    func,
+    ranges,
+    attrs=None,
+    layout=None,
+    image_name=None,
+    http_compressor="deflate",
+):
+    """
+    Apply a user defined function to an array asynchronous
+
+    **Example**
+    >>> import tiledb, tiledb.cloud, numpy
+    >>> def median(df):
+    ...   return numpy.median(df["a"])
+    >>> # Open the array then run the UDF
+    >>> tiledb.cloud.array.apply_async("tiledb://TileDB-Inc/quickstart_dense", median, [(0,5), (0,5)], attrs=["a", "b", "c"]).get()
+    2.0
+
+    :param func: user function to run
```

Comment:
Params are usually before examples (not sure if the docstring parser will even work in this order)

Reply:
I will update this and a few other places in a separate PR now